src/dict.js

```diff
@@ -157,47 +185,95 @@ Sk.builtin.dict.prototype.sq$contains = function (ob) {
 };
 
 Sk.builtin.dict.prototype.mp$ass_subscript = function (key, w) {
-    var k = kf(key);
-    var bucket = this.buckets[k.v];
-    var item;
+    const hash = kf(key);
+    let item;
+    if (typeof hash === "string") {
+        // we have a string so add to entries directly
+        item = this.entries[hash];
+        if (item === undefined) {
+            this.entries[hash] = { lhs: key, rhs: w };
+            this.size += 1;
+        } else {
+            item.rhs = w;
+        }
+    } else {
+        item = this.get$bucket_item(key, hash);
+        if (item === undefined) {
+            this.set$bucket_item(key, w, hash);
+            this.size += 1;
+        } else {
+            item.rhs = w;
+        }
+    }
+};
 
+Sk.builtin.dict.prototype.set$bucket_item = function (key, value, hash_value) {
+    // put a key, value pair into appropriate bucket and insert into entries
+    let key_hash,
+        bucket = this.buckets[hash_value];
+    const item = { lhs: key, rhs: value };
     if (bucket === undefined) {
-        // New bucket
-        bucket = {$hash: k, items: [
-            {lhs: key, rhs: w}
-        ]};
-        this.buckets[k.v] = bucket;
-        this.size += 1;
-        return;
+        bucket = this.buckets[hash_value] = [item];
+        key_hash = "#" + hash_value + "_" + 0; // this is the zeroth entry
+    } else {
+        // we might have a freeslot from deleting an item
+        // so either insert into freeslot or push
+        const free_slot_idx = bucket.indexOf(undefined);
+        if (free_slot_idx !== -1) {
+            key_hash = "#" + hash_value + "_" + free_slot_idx;
+            bucket[free_slot_idx] = item;
```

Comment:
having deleted a slot rather than set it to `undefined` this check now becomes useless (I think).
```javascript
new Array(4).indexOf(undefined) // -1
```
cleaner then just to disregard freeslots here? - I can't think of a nice way to find an empty slot. 
So either 
A - skip this logic, 
B - revert deleting the element and set it to undefined, 
C - is there a nice way to find an empty slot.

Reply:
You're right, I think B reverting deleting the element is your best option, it was a bad suggestion of mine.